libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -1860,48 +1855,52 @@ void GCS::send_textv(MAV_SEVERITY severity, const char *fmt, va_list arg_list, u
 /*
     send a statustext message to specific MAVLink connections in a bitmask
  */
-void GCS::service_statustext(void)
+void GCS_MAVLINK::service_statustext(void)
 {
-    // create bitmask of what mavlink ports we should send this text to.
-    // note, if sending to all ports, we only need to store the bitmask for each and the string only once.
-    // once we send over a link, clear the port but other busy ports bit may stay allowing for faster links
-    // to clear the bit and send quickly but slower links to still store the string. Regardless of mixed
-    // bitrates of ports, a maximum of _status_capacity strings can be buffered. Downside
-    // is if you have a super slow link mixed with a faster port, if there are _status_capacity
-    // strings in the slow queue then the next item can not be queued for the faster link
+    GCS::StatusTextQueue &_statustext_queue = gcs().statustext_queue();
+    WITH_SEMAPHORE(_statustext_queue.semaphore());
 
     if (_statustext_queue.empty()) {
         // nothing to do
         return;
     }
 
-    for (uint8_t idx=0; idx<_status_capacity; ) {
-        statustext_t *statustext = _statustext_queue[idx];
+    const uint8_t chan_bit = (1U<<chan);
+    // note the lack of idx++ here.  We may remove the iteration item
+    // from the queue as the last thing we do, in which case we don't
+    // want to move idx.
+    const uint16_t payload_size = PAYLOAD_SIZE(chan, STATUSTEXT);
+    for (uint8_t idx=0; idx<_statustext_queue.available(); ) {
+        if (txspace() < payload_size) {
+            break;
```

Comment:
It was actually the "doing work after the for loop" case I was worried about here.  If someone added more stuff into the end of this function that could be executed even with a full channel it would start getting skipped when we filled the channel, which would not be an obvious bug.

You know I'm usually a fan of early return :-)

Should I change this to just return?

Reply:
You're right Peter. Probably better to leave as a break for maintainability